Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp

```diff
@@ -0,0 +1,557 @@
+
+#if ENABLE(GPU_PROCESS) && ENABLE(VIDEO)
+
+#include "AudioVideoRendererRemoteMessageReceiverMessages.h"
+#include "GPUConnectionToWebProcess.h"
+#include "GPUProcess.h"
+#include "LayerHostingContext.h"
+#include "Logging.h"
+#include "RemoteAudioVideoRendererProxyManagerMessages.h"
+#include "RemoteVideoFrameObjectHeap.h"
+#if USE(AVFOUNDATION)
+#include <WebCore/AudioVideoRendererAVFObjC.h>
+#else
+#include <WebCore/AudioVideoRenderer.h>
+#endif
+#include <WebCore/MediaSamplesBlock.h>
+#include <WebCore/PlatformLayer.h>
+
+#include <wtf/LoggerHelper.h>
+#if PLATFORM(COCOA)
+#include <wtf/MachSendRightAnnotated.h>
+#endif
+#include <wtf/TZoneMallocInlines.h>
+
+#define MESSAGE_CHECK(assertion) MESSAGE_CHECK_BASE(assertion, m_gpuConnectionToWebProcess.get()->connection())
+#define MESSAGE_CHECK_COMPLETION(assertion, completion) MESSAGE_CHECK_COMPLETION_BASE(assertion, m_gpuConnectionToWebProcess.get()->connection(), completion)
+
+namespace WebCore {
+
+}
+
+namespace WebKit {
+
+using namespace WebCore;
+
+WTF_MAKE_TZONE_ALLOCATED_IMPL(RemoteAudioVideoRendererProxyManager);
+
+RefPtr<AudioVideoRenderer> RemoteAudioVideoRendererProxyManager::createRenderer()
+{
+#if USE(AVFOUNDATION)
+    return AudioVideoRendererAVFObjC::create(Ref { m_gpuConnectionToWebProcess.get()->logger() }, LoggerHelper::uniqueLogIdentifier());
+#else
+    ASSERT_NOT_REACHED();
+    return nullptr;
+#endif
+}
+
+RemoteAudioVideoRendererProxyManager::RemoteAudioVideoRendererProxyManager(GPUConnectionToWebProcess& connection)
+    : m_videoFrameObjectHeap(connection.videoFrameObjectHeap())
+    , m_gpuConnectionToWebProcess(connection)
+#if !RELEASE_LOG_DISABLED
+    , m_logIdentifier { LoggerHelper::uniqueLogIdentifier() }
+    , m_logger { connection.logger() }
+#endif
+{
+}
+
+RemoteAudioVideoRendererProxyManager::~RemoteAudioVideoRendererProxyManager() = default;
+
+void RemoteAudioVideoRendererProxyManager::ref() const
+{
+    m_gpuConnectionToWebProcess.get()->ref();
+}
+
+void RemoteAudioVideoRendererProxyManager::deref() const
+{
+    m_gpuConnectionToWebProcess.get()->deref();
+}
+
+std::optional<SharedPreferencesForWebProcess> RemoteAudioVideoRendererProxyManager::sharedPreferencesForWebProcess() const
+{
+    if (RefPtr gpuConnectionToWebProcess = m_gpuConnectionToWebProcess.get())
+        return gpuConnectionToWebProcess->sharedPreferencesForWebProcess();
+
+    return std::nullopt;
+}
+
+void RemoteAudioVideoRendererProxyManager::create(RemoteAudioVideoRendererIdentifier identifier)
+{
+    MESSAGE_CHECK(!m_renderers.contains(identifier));
+
+    RefPtr renderer = createRenderer();
+    ASSERT(renderer);
+    if (!renderer)
+        return;
+    RendererContext context {
+        .renderer = renderer.releaseNonNull()
+    };
+
+    context.renderer->notifyWhenErrorOccurs([weakThis = WeakPtr { *this }, identifier](PlatformMediaError error) {
+        if (RefPtr protectedThis = weakThis.get(); protectedThis && protectedThis->m_renderers.contains(identifier))
```

Comment:
This pattern of `protectedThis->m_renderers.contains(identifier)` happens a lot. Maybe you could add a convenience method like:

```cpp
private:
    bool has(RemoteAudioVideoRendererIdentifier identifier) const { return m_renderers.contains(identifier); }
```

That would turn this call site (and others) into:


```suggestion
        if (RefPtr protectedThis = weakThis.get(); protectedThis && protectedThis->has(identifier))
```